bpfd/src/bpf.rs

```diff
@@ -555,66 +557,133 @@ impl BpfManager {
+                                    priority: p.info.metadata.priority,
+                                    proceed_on: p.info.proceed_on.clone(),
+                                    direction: p.direction,
+                                    position: p.info.current_position.unwrap_or_default() as i32,
+                                },
+                            ),
                         },
                     ),
-                },
-                Program::Tc(p) => ProgramInfo {
-                    id: *id,
-                    name: p.data.section_name.to_string(),
-                    location: p.data.location.clone(),
-                    program_type: ProgramType::Tc as i32,
-                    attach_info: crate::command::AttachInfo::Tc(crate::command::TcAttachInfo {
-                        iface: p.info.if_name.to_string(),
-                        priority: p.info.metadata.priority,
-                        proceed_on: p.info.proceed_on.clone(),
-                        direction: p.direction,
-                        position: p.info.current_position.unwrap_or_default() as i32,
-                    }),
-                },
-                Program::Uprobe(p) => ProgramInfo {
-                    id: *id,
-                    name: p.data.section_name.to_string(),
-                    location: p.data.location.clone(),
-                    program_type: ProgramType::Kprobe as i32,
-                    attach_info: crate::command::AttachInfo::Uprobe(
-                        crate::command::UprobeAttachInfo {
-                            fn_name: p.info.fn_name.clone(),
-                            offset: p.info.offset,
-                            target: p.info.target.clone(),
-                            pid: p.info.pid,
-                            namespace: p.info.namespace.clone(),
+                    Program::Uprobe(p) => (
+                        prog_index,
+                        ProgramInfo {
+                            id: *id,
+                            name: p.data.section_name.to_string(),
+                            location: p.data.location.clone(),
+                            program_type: ProgramType::Kprobe as i32,
+                            attach_info: crate::command::AttachInfo::Uprobe(
+                                crate::command::UprobeAttachInfo {
+                                    fn_name: p.info.fn_name.clone(),
+                                    offset: p.info.offset,
+                                    target: p.info.target.clone(),
+                                    pid: p.info.pid,
+                                    namespace: p.info.namespace.clone(),
+                                },
+                            ),
                         },
                     ),
-                },
+                }
             })
             .collect();
-        Ok(programs)
+
+        loaded_programs()
+            .map(|p| {
```

Comment:
I would collect a `KernelProgramInfo` after we call `Load` - it can be stored in `Program` since it *shouldn't* change. One day we might want to add a reconciler that checks kernel state vs. desired state as if someone else unloads a program after we've loaded it we're in trouble anyway.

Back in relation to this code, it's then as simple as doing `loaded_programs().filter()` where the predicate function de-duplicates loaded programs by their ID. This is less error-prone than matching on name and/or tag.

Reply:
+1 this was the bit I couldn't figure out originally and at the time I was trying NOT to store a `KernelProgramInfo` in Program.

Reply:
Additionally you had asked me to use `filter` here on `loaded_programs`  with the kernel_IDs of the programs being handled by kernel ids however using a map lookup to track the bpfd programs seemed more performant since it required only iterating through the bpfd Programs and loaded programs once 

While using filter to de-duplicate based on a vector of map IDs would have required us to iterate through both lists multiple times. 

Again please point it out if I'm missing something here as I very well may be.